Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+                                title: filename,
+                                key: key,
+                                libraryId: libraryId
+                            )
+
+                            return (attachment, url)
+                        }
+
+                        return (item, attachments)
+                    } catch let error {
+                        DDLogError("IdentifierLookupController: can't parse data - \(error)")
+                        return nil
+                    }
+                }
+                
+                func process(identifier: String, response: ItemResponse, attachments: [(Attachment, URL)], libraryId: LibraryIdentifier, collectionKeys: Set<String>) {
+                    backgroundQueue.async { [weak self] in
+                        guard let self = self else { return }
+                        do {
+                            try storeDataAndDownloadAttachmentIfNecessary(identifier: identifier, response: response, attachments: attachments)
+                        } catch let error {
+                            DDLogError("IdentifierLookupController: can't create item(s) - \(error)")
+                            self.changeLookup(for: identifier, to: .failed)
+                            self.observable.on(.next(Update(kind: .itemCreationFailed(identifier: identifier, response: response, attachments: attachments))))
+                        }
+                    }
+                    
+                    func storeDataAndDownloadAttachmentIfNecessary(identifier: String, response: ItemResponse, attachments: [(Attachment, URL)]) throws {
+                        let request = CreateTranslatedItemsDbRequest(responses: [response], schemaController: schemaController, dateParser: dateParser)
+                        try dbStorage.perform(request: request, on: backgroundQueue)
+                        changeLookup(for: identifier, to: .translated(.init(response: response, attachments: attachments, libraryId: libraryId, collectionKeys: collectionKeys)))
+                        observable.on(.next(Update(kind: .itemStored(identifier: identifier, response: response, attachments: attachments))))
+                        
+                        guard Defaults.shared.shareExtensionIncludeAttachment else { return }
+
+                        let downloadData = attachments.map({ ($0, $1, response.key) })
+                        guard !downloadData.isEmpty else { return }
+                        remoteFileDownloader.download(data: downloadData)
+                        observable.on(.next(Update(kind: .pendingAttachments(identifier: identifier, response: response, attachments: attachments))))
+                    }
+                }
+            }
+        }
+    }
+    
+    // MARK: Lookup Data
+    func cleanupLookup(force: Bool, alreadyInQueue: Bool, update: Update?) {
```

Comment:
I don't love the `alreadyInQueue` parameter. You specify `_cleanupLookup` function anyway, so I'd just have separate functions, where `cleanupLookup` calls `_cleanupLookup` on proper queue and if you know you're on proper queue you can call `_cleanupLookup` directly.

I'd also change the name to `cleanupLookupIfNeeded` since you check for finished tasks in it.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/1196ff95e7aa78d9f47d4d5b321dd5dd96f0c6a4

Reply:
Didn't know you can check queue like that, nice :)